stingray/deadtime/filters.py

```diff
@@ -0,0 +1,161 @@
+    dt_sigma : float
+        The standard deviation of a non-constant dead time around deadtime.
+    return_all : bool
+        If True, return the mask that filters the input event list to obtain
+        the output event list.
+
+    Returns
+    -------
+    new_ev_list : class:`stingray.events.EventList` object
+        The filtered event list
+    additional_output : dict
+        Object with all the following attributes. Only returned if
+        `return_all` is True
+    uf_events : array-like
+        Unfiltered event list (events + background)
+    is_event : array-like
+        Boolean values; True if event, False if background
+    deadtime : array-like
+        Dead time values
+    bkg : array-like
+        The filtered background event list
+    mask : array-like, optional
+        The mask that filters the input event list and produces the output
+        event list.
+
+    """
+    additional_output = DeadtimeFilterOutput()
+    if isinstance(event_list, EventList):
+        ev_list = event_list.time
+        mjdref = event_list.mjdref
+    else:
+        ev_list = event_list
+        mjdref = 0
+
+    if deadtime <= 0.:
+        if deadtime < 0:
+            warnings.warn("Dead time < 0")
+        return event_list
+
+    # Create the total lightcurve, and a "kind" array that keeps track
+    # of the events classified as "signal" (True) and "background" (False)
+    if bkg_ev_list is not None:
+        tot_ev_list = np.append(ev_list, bkg_ev_list)
+        ev_kind = np.append(np.ones(len(ev_list), dtype=bool),
+                            np.zeros(len(bkg_ev_list), dtype=bool))
+        order = np.argsort(tot_ev_list)
+        tot_ev_list = tot_ev_list[order]
+        ev_kind = ev_kind[order]
+        del order
+    else:
+        tot_ev_list = ev_list
+        ev_kind = np.ones(len(ev_list), dtype=bool)
+
+    nevents = len(tot_ev_list)
+    all_ev_kind = ev_kind.copy()
+
+    if dt_sigma is not None:
+        deadtime_values = ra.normal(deadtime, dt_sigma, nevents)
+    else:
+        deadtime_values = np.zeros(nevents) + deadtime
+
+    initial_len = len(tot_ev_list)
+
+    if paralyzable:
+        tot_ev_list, saved_mask = \
+            _paralyzable_dead_time(tot_ev_list, deadtime_values)
+
+    else:
+        tot_ev_list, saved_mask = \
+            _non_paralyzable_dead_time(tot_ev_list, deadtime_values)
+
+    ev_kind = ev_kind[saved_mask]
+    deadtime_values = deadtime_values[saved_mask]
+    final_len = len(tot_ev_list)
+    log.info(
+        'filter_for_deadtime: '
+        '{0}/{1} events rejected'.format(initial_len - final_len,
+                                         initial_len))
+    retval = EventList(time=tot_ev_list[ev_kind], mjdref=mjdref)
+
+    if isinstance(event_list, EventList) and hasattr(event_list, 'pi') and event_list.pi is not None:
+        warnings.warn(
+            "PI information is lost during dead time filtering",
+            AstropyUserWarning)
```

Comment:
Why are we losing PI information during dead time filtering? Aren't we keeping track of which events are kept and which ones are lost, anyway, that we could apply that mask to the PI channels? Or am I missing something?

Reply:
Ok. (Just being lazy :P). Done